// @mintlify/validation/dist/index.js
import { zodToJsonSchema } from 'zod-to-json-schema';
import { colorsSchema } from './mint-config/schemas/colors.js';
import { mintConfigSchema } from './mint-config/schemas/config.js';
export * from './openapi/types/endpoint.js';
export * from './openapi/OpenApiToEndpointConverter.js';
export { stripComponents } from './openapi/stripComponents.js';
export { SchemaConverter } from './openapi/SchemaConverter.js';
export { generateExampleFromSchema } from './openapi/generateExampleFromSchema.js';
export { generateFirstIncrementalSchema, generateNextIncrementalSchema, } from './openapi/IncrementalEvaluator.js';
export { validateMintConfig } from './mint-config/validateMintConfig.js';
export { formatIssue } from './mint-config/formatIssue.js';
export const mintConfigJsonSchema = (() => {
    const cleanedSchema = mintConfigSchema.extend({
        colors: colorsSchema.omit({
            ultraDark: true,
            ultraLight: true,
        }),
    });
    return zodToJsonSchema(cleanedSchema, 'Schema');
})();
